### DISCUSSION TITLE
AST Printer API

### DISCUSSION THREAD

**Discussion by siketyan** [microsoft/typespec#4775] [Ideas]
I want an AST printer API that prints AST node(s) to string or a file. This will help creating tools like (something) to TypeSpec converter, or generating TypeSpec files from scratch.

There is an AST printer for formatting TypeSpec files with Prettier API, but it is not exported and can't be used outside the compiler:
https://github.com/microsoft/typespec/blob/main/packages/compiler/src/formatter/print/printer.ts

Is there any plan to expose this API or creating a new API for printing AST?

**Comment by timotheeguerin** [comment]
We do not expose that today, however I have a PR that added some new tool that included the need to expose it https://github.com/microsoft/typespec/pull/4383/files#diff-f1970e6a8b9b68f5eb9cb0acd29a78932e907bfbb05bbb0bd456bb506d885563

There is a few things to watch out there is still some nodes that are not formatted or by design just keep the original text(e.g. a string node) and for that the nodes compute the original text from the original document. That pr does fix some of the cases and allow passing the `rawText` as an hidden property of the node for the printer so it won't need to look at the whole file but its something to know and we haven't really confirmed if thats the way we want to move forward

**Reply by siketyan** [reply]
That is the API that I exactly wanted!

```ts
export function printTypeSpecNode(node: Node): Promise<string>;
```

Though some nodes aren't formatted, that API will the good start for the AST printer API. I'll keep watching the PR for now.

In addition to this, it will be great if there is a AST factory API like TypeScript does (`ts.factory`).

Thank you for your answer 🙏
